### PR TITLE
Update indexing material

### PR DESCRIPTION
**Pull request by negin513**
Adding two new notebooks in support of SciPy 2023: 

1. 02.1_indexing_Basic.ipynb (Comprehensive coverage of all basic indexing)
2. 02.2_indexing_Advanced.ipynb 

The first notebook includes comprehensive coverage of all basic indexing and can basically replace the 02.1_working_with_labeled_data.ipynb. But I wanted to first ask for feedback before removing the other notebook. 

The second notebook includes : 
* Vectorized Indexing
* Boolean Indexing and `isin`, and masking using `where`

Relevant Issue: https://github.com/xarray-contrib/xarray-tutorial/issues/170

We can add materials for  Fancy Datetime Indexing here @andersy005 .

### CODE REVIEW

**Comment by dcherian** [comment]
Very nice! I just took a quick look, will do a more thorough review tomorrow.

> The first notebook includes comprehensive coverage of all basic indexing and can basically replace the 02.1_working_with_labeled_data.ipynb. But I wanted to first ask for feedback before removing the other notebook.

Perhaps we can copy the exercises over? That notebook also has some date time indexing material that could be copied?

Can you add entries in `_toc.yml` please? I think your second notebook should go in the Intermediate section.

File: fundamentals/02.1_indexing_Basic.ipynb
Comment (dcherian) [inline fundamentals/02.1_indexing_Basic.ipynb:1]: - [ ] Add to `_toc.yml` replacing the current notebook. 
- [ ] Please set up a redirect in `_config.yml`
- [ ] Update the link under "Fundamental learning path"

Comment (dcherian) [reply fundamentals/02.1_indexing_Basic.ipynb:1]: "Indexing and Selecting Data" is repeated twice.  

We can delete the Xarray logo at the top.

Comment (dcherian) [reply fundamentals/02.1_indexing_Basic.ipynb:1]: I would add "Understanding the difference between position and label-based indexing" under learning goals

Comment (dcherian) [reply fundamentals/02.1_indexing_Basic.ipynb:1]: We can replace the `div`s with [myst syntax](https://myst-parser.readthedocs.io/en/latest/syntax/admonitions.html#admonitions).

Comment (dcherian) [reply fundamentals/02.1_indexing_Basic.ipynb:1]: "all of these methods work on Dataset too" -> make it a `tip`

Comment (dcherian) [reply fundamentals/02.1_indexing_Basic.ipynb:1]: I suggest minimizing references to pandas. It's common complaint that it makes things more confusing.

Comment (dcherian) [reply fundamentals/02.1_indexing_Basic.ipynb:1]: Let's copy over the exercises from https://tutorial.xarray.dev/fundamentals/02.1_working_with_labeled_data.html

We'll have to migrate to the new sphinx-exercise syntax:
`````
```{exercise}
:label: indexing-1
Text
```

````{solution} indexing-1
:class: dropdown

```python
solution_code
```
````

`````

File: fundamentals/02.2_indexing_Advanced.ipynb
Comment (dcherian) [inline fundamentals/02.2_indexing_Advanced.ipynb:1]: remove the logo.

Comment (dcherian) [reply fundamentals/02.2_indexing_Advanced.ipynb:1]: I think you'll need one or two more lines to explain "orthogonal" indexing.

We should also link to the xarray docs appropriately.

Comment (dcherian) [reply fundamentals/02.2_indexing_Advanced.ipynb:1]: Let's add a new subsection for "pointwise indexing"

Comment (dcherian) [reply fundamentals/02.2_indexing_Advanced.ipynb:1]: Let's use the myst syntax where we can instead of alerts and bold font

Comment (dcherian) [reply fundamentals/02.2_indexing_Advanced.ipynb:1]: Masking and `isin` can be grouped under "Boolean indexing". They could be moved to its own notebook.

Comment (dcherian) [reply fundamentals/02.2_indexing_Advanced.ipynb:1]: We can skip "align and reindex". Reindexing can be its own section left for some future work.

**Comment by dcherian** [comment]
Here's a proposal for organizing:

- Fundamentals
  - Labeled data
    - Selecting Data <- First notebook : positional vs label-based indexing
    - Selection with Times <- Datetime Indexing stuff
 - Intermediate
   - Indexing 
     - Vectorized and Pointwise indexing
     - Boolean Indexing & Masking

What do you think? We've been evolving towards breaking up the material in to small pieces so it's easy to choose subsets for a syllabus.

File: fundamentals/02.1_indexing_Basic.ipynb
Comment (negin513) [reply fundamentals/02.1_indexing_Basic.ipynb:1]: Thanks for your comments @dcherian. I think I have addressed all these comments. I am not sure if I am setting up the redirect correctly in `_config.yml`.

File: fundamentals/02.2_indexing_Advanced.ipynb
Comment (negin513) [reply fundamentals/02.2_indexing_Advanced.ipynb:1]: Thanks for your comments. I significantly changed this notebook based on your comments. 
I added a section called orthogonal indexing. I tried to add the pointwise terminology as you recommended. 
For the boolean masking :
I created a seperate notebook for Boolean masking/indexing (as you recommended). 
I updated the notebook to use `rasm` dataset and created a specific section/exercise telling them why they cannot use `sel` for such datasets for selecting regions.

**Comment by negin513** [comment]
> Here's a proposal for organizing:
> 
> * Fundamentals
>   
>   * Labeled data
>     
>     * Selecting Data <- First notebook : positional vs label-based indexing
>     * Selection with Times <- Datetime Indexing stuff
> * Intermediate
>   
>   * Indexing
>     
>     * Vectorized and Pointwise indexing
>     * Boolean Indexing & Masking
> 
> What do you think? We've been evolving towards breaking up the material in to small pieces so it's easy to choose subsets for a syllabus.
> 
> We'll need to add the parts you want to cover to the `workshops/scipy2023/README.md` in the appropriate dropdown.

Thanks @dcherian for your recommendation. As you suggested I created a notebook for Boolean Indexing & Masking.

**Comment by negin513** [comment]
@andersy005, please feel free to add a section to the basic or advanced indexing notebook on the fancy datetime indexing. I think at this point I am not going to make any other changes to these notebooks unless I hear back from you or @dcherian .

**Comment by negin513** [comment]
@dcherian I am not sure why the JupyterBuild of this notebook fails:
It complains about the redirect and a page already existing:
``` WARNING: (broken) fundamentals/02.1_working_with_labeled_data.html redirects to fundamentals/02.1_indexing_Basic.html but /home/runner/work/xarray-tutorial/xarray-tutorial/_build/html/fundamentals/02.1_working_with_labeled_data.html already exists!
```

Should I manually deleted the cached htmls ?

**Comment by dcherian** [comment]
You can change the key to force a new cache:
https://github.com/xarray-contrib/xarray-tutorial/blob/2ca47a9a5a1402e0035dec423e3c1d7a880e036e/.github/workflows/main.yaml#L32-L33

**Comment by andersy005** [comment]
> You can change the key to force a new cache:
> 
> https://github.com/xarray-contrib/xarray-tutorial/blob/2ca47a9a5a1402e0035dec423e3c1d7a880e036e/.github/workflows/main.yaml#L32-L33

i usually delete stale cache via the actions UI, and this seems to invalidate it and force a new cache (under the same key though) 


Screenshot 2023-07-07 at 21 35 20

**Comment by andersy005** [comment]
> @andersy005, please feel free to add a section to the basic or advanced indexing notebook on the fancy datetime indexing. I think at this point I am not going to make any other changes to these notebooks unless I hear back from you or @dcherian .

thank you for putting this together, @negin513! i intend to push changes to this PR tomorrow and sunday

File: fundamentals/02.1_indexing_Basic.ipynb
Comment (andersy005) [inline fundamentals/02.1_indexing_Basic.ipynb:1]: @negin513 / @dcherian, i just added a section on datetime index to this notebook, let me know how this looks

**Comment by dcherian** [comment]
Thanks all. This is looking great.

 I pushed some minor edits. 

Some major comments:
- let's delete the numpy advanced indexing piece from "basic indexing" material. That's going to be confusing. It can be moved to the "advanced indexing" notebook if you prefer.
- "basic indexing" uses datetime indexing before it is introduced. Let's edit that to select along `lat` instead of `time`. or reorder the material

**Comment by dcherian** [comment]
We only have 30 minutes for the indexing topic, so perhaps just do only the Advanced and Vectorized Indexing notebook?

Given that I'm just going to merge here. 

Thanks @negin513 and @andersy005 !

**Comment by negin513** [comment]
Sounds great. Thanks @dcherian. Do we have a schedule now and how long each topic is going to take?

**Pull request merged**